api/marketdata.go

```diff
@@ -156,12 +156,40 @@ func getChartsHandler(charts *market.Charts) func(c *gin.Context) {
 			ginutils.RenderError(c, http.StatusBadRequest, "Invalid time_start provided")
 			return
 		}
-		maxItems, err := strconv.Atoi(c.Query("max_items"))
+		maxItemsRaw := c.Query("max_items")
+		maxItems, err := strconv.Atoi(maxItemsRaw)
 		if err != nil || maxItems <= 0 {
 			maxItems = defaultMaxChartItems
 		}
 
 		currency := c.DefaultQuery("currency", watchmarket.DefaultCurrency)
+
+		var isValidTimestamp bool
+		var response watchmarket.ChartData
+
+		key := cache.GenerateKey(coinQuery + token + maxItemsRaw + currency)
+		cacheData, err := cache.DB.Get(key)
+		if err == nil {
+			isValidTimestamp = cacheData.Validate(timeStart, minute*5)
+		} else if err != storage.ErrNotExist {
+			logger.Warn("Failed to Get Cache", logger.Params{"err": err.Error()})
+		}
+
+		if isValidTimestamp {
+			err = json.Unmarshal(cacheData.RawData, &response)
+			if err == nil {
+				ginutils.RenderSuccess(c, response)
+				return
+			}
+		}
+
+		if err == nil {
+			deleteResult, err := cache.DB.Delete(key)
+			if err != nil || deleteResult != storage.SaveResultSuccess {
+				logger.Warn("Failed to Delete Cache", logger.Params{"err": err})
+			}
+		}
```

Comment:
this is too much logic for the controller. please encapsulate elsewhere

Reply:
critical: you didn't test this in tests.

Reply:
There are tests (integration, also I cannot see any possibility to encapsulate it, it is tightly linked to the controller logic)